### PR TITLE
Feature/crud for board

### PR DESCRIPTION
**Pull request by KurnosovaElena**

### CODE REVIEW

File: Management_System/BLL/BLL.csproj [hunk 1]
```diff
@@ -6,6 +6,10 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.10" />
```
Comment (Vlada25) [inline Management_System/BLL/BLL.csproj]: You don't need in EntityFrameworkCore here as you've already have it in DAL

File: Management_System/DAL/DI/ServicesConfguration.cs [hunk 2]
```diff
@@ -11,5 +13,7 @@ public static class ServicesConfguration
     public static void AddDataAccessDependencies(this IServiceCollection services, IConfiguration configuration)
     {
         services.AddDbContext<ManagementSystemDBContext>(options => options.UseNpgsql(configuration.GetConnectionString(Connections.DBConnection)));
+
+        services.AddTransient<IBoardRepository, BoardRepository>();
```
Comment (Vlada25) [inline Management_System/DAL/DI/ServicesConfguration.cs]: Why transient? (You configured BoardService as Scoped, so seems like repo also shoud be configured as scoped)

**Review by DmitriyG0r4anin — APPROVED** [review]

File: Management_System/BLL/Services/Implementaiton/BoardService.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using DAL.Repositories.Interfaces;
+
+namespace BLL.Services.Implementaiton;
+public class BoardService(IBoardRepository repository) : IBoardService
+{
+    public async Task<BoardEntity> Add(BoardEntity entity, CancellationToken cancellationToken)
+    {
+        await repository.Add(entity, cancellationToken);
+
+        return entity;
```
Comment (5Ka-me) [inline Management_System/BLL/Services/Implementaiton/BoardService.cs:13]: Is the instance of entity will be updated?

File: Management_System/BLL/Services/Implementaiton/BoardService.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        return boards;
+    }
+
+    public async Task<BoardEntity> Update(Guid id, BoardEntity entity, CancellationToken cancellationToken)
+    {
+        var board = await repository.GetBoard(id, cancellationToken);
+
+        await repository.Update(board, cancellationToken);
+
+        return entity;
```
Comment (5Ka-me) [inline Management_System/BLL/Services/Implementaiton/BoardService.cs:35]: Same about the updated entity

File: Management_System/DAL/Repositories/Implementations/BoardRepository.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+using DAL.Repositories.Interfaces;
+using Microsoft.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DAL.Repositories.Implementations;
+public class BoardRepository (ManagementSystemDBContext db) : Repository<BoardEntity>(db), IBoardRepository
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Implementations/BoardRepository.cs]: Add blank line please

File: Management_System/DAL/Repositories/Implementations/Repository.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+﻿using DAL.Context;
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Implementations/Repository.cs:1]: How about to add `Get(Guid id)` here?

File: Management_System/DAL/Repositories/Interfaces/IRepository.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿namespace DAL.Repositories.Interfaces;
+
+public interface IRepository<T> where T : class
+{
+    IQueryable<T> GetAll(CancellationToken cancellationToken);
+    IQueryable<T> GetAll(int page, int pageSize, CancellationToken cancellationToken);
+    Task Add(T entity, CancellationToken cancellationToken);
+    Task Update(T entity, CancellationToken cancellationToken);
+    Task Delete(T entity, CancellationToken cancellationToken);
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Interfaces/IRepository.cs]: Why not to include _Get single_?

File: Management_System/DAL/Entities/BoardEntity.cs [hunk 8]
```diff
@@ -7,6 +7,6 @@ public class BoardEntity
     public string Name { get; set; } = string.Empty;
     public string Description { get; set; } = string.Empty;
 
-    public IEnumerable<TableEntity> Tables { get; set; } = [];
-    public IEnumerable<UserBoardEntity> UserBoards { get; set; } = [];
+    public IEnumerable<TableEntity> Tables { get; set; } = new List<TableEntity>();
+    public IEnumerable<UserBoardEntity> UserBoards { get; set; } = new List<UserBoardEntity>();
```
Comment (5Ka-me) [inline Management_System/DAL/Entities/BoardEntity.cs:11]: Why did you do that?

File: Management_System/DAL/Repositories/Implementations/BoardRepository.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    public async Task<IEnumerable<BoardEntity>> GetBoards(CancellationToken cancellationToken)
+    {
+        var boards = await GetAll(cancellationToken).ToListAsync(cancellationToken);
+        return boards;
+    }
+    public async Task<BoardEntity?> GetBoard(Guid id, CancellationToken cancellationToken)
+    {
+        var board = await db.Boards.Include(b => b.Tables).Include(b => b.UserBoards).FirstOrDefaultAsync(b => b.Id == id, cancellationToken);
+        return board;
+    }
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Implementations/BoardRepository.cs]: Why didn't you use existing methods like `GetAll` from generic repo?

File: Management_System/Management_System/Controllers/BoardController.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    public async Task Update(Guid id, BoardEntity entity, CancellationToken cancellationToken)
+    {
+        await service.Update(id, entity, cancellationToken);
+    }
+
+    [HttpDelete("{id}")]
+    public async Task Delete(Guid id, CancellationToken cancellationToken)
+    {
+        await service.Delete(id, cancellationToken);
+    }
```
Comment (5Ka-me) [inline Management_System/Management_System/Controllers/BoardController.cs]: There is no exceptions checks. Are you going to add them in separate PR or you just missed that or you add that with ExceptionHandlerMiddleware? 😅

File: Management_System/BLL/Services/Implementaiton/BoardService.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public async Task<BoardEntity> Delete(Guid id, CancellationToken cancellationToken)
+    {
+        var board = await repository.GetBoard(id, cancellationToken);
+
+        await repository.Delete(board, cancellationToken);
+
+        return board;
```
Comment (5Ka-me) [inline Management_System/BLL/Services/Implementaiton/BoardService.cs]: Do you really need to return deleted entity?

File: Management_System/DAL/Repositories/Interfaces/IRepository.cs [hunk 12]
```diff
@@ -0,0 +1,10 @@
+﻿namespace DAL.Repositories.Interfaces;
+
+public interface IRepository<T> where T : class
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Interfaces/IRepository.cs]: Please let do code style in one way.
Do we add blank line after namespace or not?

Fix all updated files please

File: Management_System/DAL/Repositories/Interfaces/IBoardRepository.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿using DAL.Entities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DAL.Repositories.Interfaces;
+public interface IBoardRepository : IRepository<BoardEntity>
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Interfaces/IBoardRepository.cs]: Please let do code style in one way.
Do we add blank line after namespace or not?

Fix all updated files please

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: Management_System/DAL/Entities/BoardEntity.cs (hunk 8 above)
Comment (KurnosovaElena) [reply Management_System/DAL/Entities/BoardEntity.cs:11]: I had "System.Text.Json.JsonException" when I tried to get the board by id

File: Management_System/Management_System/Controllers/BoardController.cs (hunk 10 above)
Comment (KurnosovaElena) [reply Management_System/Management_System/Controllers/BoardController.cs]: I'll do it later with ExceptionHandlerMiddleware

File: Management_System/BLL/Services/Implementaiton/BoardService.cs (hunk 3 above)
Comment (KurnosovaElena) [reply Management_System/BLL/Services/Implementaiton/BoardService.cs:13]: I'll change entity to dto later (if I get you right)

Comment (5Ka-me) [reply Management_System/BLL/Services/Implementaiton/BoardService.cs:13]: I mean that at the beginning of the method you have entity without id. This field will be set up by DB. So, don't you need to have full entity after the query has finished?

File: Management_System/DAL/Entities/BoardEntity.cs (hunk 8 above)
Comment (5Ka-me) [reply Management_System/DAL/Entities/BoardEntity.cs:11]: How it is related to the selected lines?

File: Management_System/BLL/Services/Interfaсes/IBoardService.cs [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+namespace BLL.Services.Interfaсes;
+
+public interface IBoardService
+{
+    Task<BoardEntity> GetById(Guid id, CancellationToken cancellationToken);
+    Task<IEnumerable<BoardEntity>> GetAll(CancellationToken cancellationToken);
+    Task<BoardEntity> Add(BoardEntity entity, CancellationToken cancellationToken);
+    Task<BoardEntity> Update(Guid id, BoardEntity entity, CancellationToken cancellationToken);
+    Task Delete(Guid id, CancellationToken cancellationToken);
+}
```
Comment (5Ka-me) [inline Management_System/BLL/Services/Interfaсes/IBoardService.cs:17]: Perhaps it is better to move this to a generic service.
You can do this later if you want to

File: Management_System/DAL/Repositories/Implementations/Repository.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+                     .Take(pageSize)
+                     .AsNoTracking();
+        }
+
+        public IQueryable<T> GetById(Expression<Func<T, bool>> predicate)
+        {
+            return db.Set<T>()
+                .Where(predicate)
+                .AsNoTracking();
+        }
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Implementations/Repository.cs]: Are you sure that this predicate is a _GetById_?

File: Management_System/DAL/Repositories/Implementations/Repository.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            await db.AddAsync(entity, cancellationToken);
+            await db.SaveChangesAsync(cancellationToken);
+        }
+
+        public async Task Update(T entity, CancellationToken cancellationToken)
+        {
+            db.Update(entity);
+            await db.SaveChangesAsync(cancellationToken);
+        }
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Implementations/Repository.cs]: Same as in one of my previous comments.
Why you don't return updated entities?

File: Management_System/DAL/Repositories/Interfaces/IBoardRepository.cs [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DAL.Repositories.Interfaces;
+
+public interface IBoardRepository : IRepository<BoardEntity>
+{
+    Task<IEnumerable<BoardEntity>> GetBoards(CancellationToken cancellationToken);
```
Comment (5Ka-me) [inline Management_System/DAL/Repositories/Interfaces/IBoardRepository.cs]: How is it different from _GetAll_ in _IRepository_?

File: Management_System/Management_System/Controllers/BoardController.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+        var board = await service.Add(entity, cancellationToken);
+
+        return board;
+    }
+
+    [HttpPut("{id}")]
+    public async Task Update(Guid id, BoardEntity entity, CancellationToken cancellationToken)
+    {
+        await service.Update(id, entity, cancellationToken);
+    }
```
Comment (5Ka-me) [inline Management_System/Management_System/Controllers/BoardController.cs]: Why don't return entity as in the _Add_ method?

**Review by 5Ka-me — CHANGES_REQUESTED** [review]

File: Management_System/BLL/Services/Interfaсes/IBoardService.cs (hunk 14 above)
Comment (KurnosovaElena) [reply Management_System/BLL/Services/Interfaсes/IBoardService.cs:17]: I'll do it later

**Review by 5Ka-me — APPROVED** [review]

**Pull request merged**